### DISCUSSION TITLE
MODBUS for RS485 Soil Sensor

### DISCUSSION THREAD

**Discussion by garudaonekh** [arendst/Tasmota#12704] [Ideas]
Hi,
Looking for way to integrate several soil moisture sensors. It's RS485. It's multi parameter sensors. Ex. To read the 
the Phosphorous of the soil, we need to send this command :{0x01,0x03, 0x00, 0x1f, 0x00, 0x01, 0xb5, 0xcc}. I think TCP Serial bridge maybe the right path.

My problem is: 
1. Integrate it into Tasmota
2. Allow multiple sensors in a single esp32 Tasmota. We need around 6 soil moisture sensors per greenhouse.
 
It's exactly the same as this Arduino tutorial.
https://how2electronics.com/measure-soil-nutrient-using-arduino-soil-npk-sensor/

**Comment by barbudor** [comment]
Hi
The RS485 converter (raw max485) shown in the article is not supported by Tasmota (Tasmota do not manage DE and RE pins). You must use the supported converter which manage bus direction automatically. Search in the discussions or in the docs as this has already been discussed for smart meters.

Modbus protocole is supported with Scripting Language but I think this is only for smart meters. Not sure it can be used for a soil sensor.

Not Modbus is a bus so if all your sensors can have a different address, then they can all be connected to the same bus. 

If you go for TCP Serial bridge, tasmota is just a socket to UART bridge. It doesn't make use of MQTT which means you will need a computer software that will discuss through a socket with the sensors. Tasmota will just act as a bidirectional pass-through (ie bridge).

If you want to use MQTT you should use SerialBridge. 

And finally have you looked at the already supported sensors such as the Adafruit seesaw based one (i2c)?

**Reply by garudaonekh** [reply]
Yes, I have the MAX485 that supported by Tasmota. I have used several i2C sensors but when it comes to long distance, we need to use RS485.

I have my controller software which will talk with Tasmota (ESP32) thru HTTP or MQTT so I can send the command to read and parse the response back to me so what I want is for Tasmota to forward the request thru serial to the sensors.

**Reply by barbudor** [reply]
So look at SerialBride with commands SerialSend\<x\>
Received data will trigger a SerialReceived MQTT message

**Reply by garudaonekh** [reply]
Thanks, Works like a charm!! My next task is to allow multiple devices on the same RS485. It looks like there's no method to do here. Maybe I need to use relay to switch one device on while keep the other off.

**Reply by barbudor** [reply]
If the device is MODBUS compliant you should be able to set an different address for each and have them all connected in parallel on the bus. The address should be part of the message. Very similar to i2c.

**Reply by barbudor** [reply]
On the manufacturer site there is a diagram with multiple devices connected in parallel. So there must be a way to change the device's address. It could be a dip switch or some jumpers I side.
I couldn't find a doc/manual on the site. 
Address seems to be the 1st byte of the message (0x01 in your examples).
